elixir/apps/domain/lib/domain/tokens.ex

```diff
@@ -0,0 +1,211 @@
+         queryable =
+           Token.Query.by_id(id)
+           |> Token.Query.by_account_id(account_id)
+           |> Token.Query.by_type(context.type)
+           |> Token.Query.not_expired(),
+         {:ok, token} <- Repo.fetch(queryable),
+         true <-
+           Domain.Crypto.equal?(:sha, nonce <> secret <> token.secret_salt, token.secret_hash) do
+      Token.Changeset.use(token, context)
+      |> Repo.update()
+    else
+      {:error, :invalid} -> {:error, :invalid_or_expired_token}
+      {:ok, _token_payload} -> {:error, :invalid_or_expired_token}
+      {:error, :not_found} -> {:error, :invalid_or_expired_token}
+      false -> {:error, :invalid_or_expired_token}
+      _other -> {:error, :invalid_or_expired_token}
+    end
+  end
+
+  def peek_token(encoded_token, %Auth.Context{} = context) do
+    with [nonce, encoded_fragment] <- String.split(encoded_token, ".", parts: 2),
+         {:ok, {account_id, id, secret}} <- verify_token(encoded_fragment, context) do
+      {:ok, {account_id, id, nonce, secret}}
+    end
+  end
+
+  defp verify_token(encrypted_token, %Auth.Context{} = context) do
+    config = fetch_config!()
+    key_base = Keyword.fetch!(config, :key_base)
+    shared_salt = Keyword.fetch!(config, :salt)
+    salt = shared_salt <> to_string(context.type)
+
+    Plug.Crypto.verify(key_base, salt, encrypted_token, max_age: :infinity)
+  end
+
+  def delete_token(%Token{} = token, %Auth.Subject{} = subject) do
+    required_permissions =
+      {:one_of,
+       [
+         Authorizer.manage_tokens_permission(),
+         Authorizer.manage_own_tokens_permission()
+       ]}
+
+    with :ok <- Auth.ensure_has_permissions(subject, required_permissions) do
+      Token.Query.by_id(token.id)
+      |> Authorizer.for_subject(subject)
+      |> Repo.fetch_and_update(with: &Token.Changeset.delete/1)
+      |> case do
+        {:ok, token} ->
+          Phoenix.PubSub.broadcast(Domain.PubSub, "sessions:#{token.id}", "disconnect")
+          {:ok, token}
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+    end
+  end
+
+  def delete_tokens_for(%Auth.Subject{} = subject) do
+    with :ok <- Auth.ensure_has_permissions(subject, Authorizer.manage_own_tokens_permission()) do
+      Token.Query.by_actor_id(subject.actor.id)
+      |> Authorizer.for_subject(subject)
+      |> delete_tokens()
+    end
+  end
+
+  def delete_tokens_for(%Actors.Actor{} = actor, %Auth.Subject{} = subject) do
+    with :ok <- Auth.ensure_has_permissions(subject, Authorizer.manage_tokens_permission()) do
+      Token.Query.by_actor_id(actor.id)
+      |> Authorizer.for_subject(subject)
+      |> delete_tokens()
+    end
+  end
+
+  def delete_expired_tokens do
+    Token.Query.expired()
+    |> delete_tokens()
+  end
+
+  defp delete_tokens(queryable) do
+    {count, ids} =
+      queryable
+      |> Ecto.Query.select([tokens: tokens], tokens.id)
+      |> Repo.update_all(set: [deleted_at: DateTime.utc_now()])
```

Comment:
As I was thinking about this, I realized I don't have a good feel for how many token we expect per user/client.  Do we risk creating a really large table by soft deleting here?

Reply:
@bmanifold we kinda have to create a record per session so that we can revoke them and terminate all sessions, don't think there is a good way around that. Yes, the tokens table will be pretty large.

Reply:
Gotcha.  Yeah, I figured we had to create a token per session, but was just wondering if we might want to do hard deletes for the tokens rather than soft deletes so that table doesn't get really huge.  If that breaks other things, that's understandable, but was just curious.

Reply:
@bmanifold We will use partial indexes so soft deletion is not a big deal for queries we are doing and Postgres in general, if we hit a problem with it we can always hard delete tokens after some retention period.